### PR TITLE
[WIP] Feature to allow setting FQDN as hostname of pods.

### LINKED ISSUE #91036: Allow setting FQDN as hostname of pods.
**What would you like to be added**:

The ability to set the Linux kernel hostname field ([the nodename field of struct utsname](http://man7.org/linux/man-pages/man2/uname.2.html)) of a pod to its fully qualified domain name (FQDN). Currently, when a pod is configured to have FQDN (e.g., foo.subdomain.namespace.svc.company.com), the hostname field of the kernel includes its shortname. Hence, uname -n returns foo, and hostname --fqdn returns the FQDN (foo.subdomain.namespace.svc.company.com). When this new feature is enabled, both, uname -n and hostname --fqdn would return the pod’s FQDN. The shortname can still be obtained using hostname --short. 

**Why is this needed**:

The main reason for this feature is to increase the interoperability of Kubernetes. Traditionally, Unix and certain Linux distributions, such as [RedHat ](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/networking_guide/ch-configure_host_names) and CentOS, have recommended setting the kernel hostname field to the host FQDN. As a consequence, many applications created before Kubernetes rely on this behavior. Some use cases include validating licencies, self-registration of the host where an application runs in a service registry for service discovery, and obtaining the datacenter by parsing the hostname. 

**Current Approach**
<del>We have created the pull request [#91035](https://github.com/kubernetes/kubernetes/pull/91035) with this approach. We believe that Kubelet is a good place to enable this feature as it is a system configuration that should be configured by the administrators of the cluster. In our company, the assumption is that the FQDN is set as hostname and having the control over the feature allows us to be consistent across platforms as well as to avoid users’ mistakes that we would have to help troubleshoot.</del>

_UPDATE_: The PR [#91699](https://github.com/kubernetes/kubernetes/pull/91699) has been changed and tested to use a PodSpec field as described in approved KEP (#1792](https://github.com/kubernetes/enhancements/pull/1792)

Since Kernel hostname is limited to 64bytes, there is a user-facing failure scenario. In the current approach when a pod FQDN is longer than 64bytes, the pods will stay in ContainerCreating state logging errors regarding the size of FQDN. 

We are happy to lead the change through all its phases, and we will create a KEP once we reach a consensus on the approach.

/kind feature
/sig node
/sig network

### PR DESCRIPTION
**Pull request by javidiaz**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This PR allows to configure pods such that it will set the kernel hostname field
(the nodename field of struct utsname) of a pod to its full qualified domain name (FQDN). 
If the pod does not have FQDN then this has no effect.

**Which issue(s) this PR fixes**:

Fixes #91036

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
Yes, when feature is enabled.

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
 [KEP]: https://github.com/kubernetes/enhancements/pull/1792

```docs

```

/sig node
/sig network
/cc @thockin

### CODE REVIEW

**Comment by javidiaz** [comment]
I signed it

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -435,6 +435,21 @@ func (kl *Kubelet) GetPodCgroupParent(pod *v1.Pod) string {
 	return cgroupParent
 }
 
+// getHostNameForKernel gets hostname value to set in the hostname field (the nodename field of struct utsname) of the pod.
```
Comment (dcbw) [inline pkg/kubelet/kubelet_pods.go:438]: Code comment function name must match actual function name

**Comment by dcbw** [comment]
Does this work the right way on Windows too?

**Comment by javidiaz** [comment]
> Does this work the right way on Windows too?
We did some basic test with docker -h option and Docker just set whatever string you send to the COMPUTERNAME env so hostname returns it.

> docker run -h "test.company.com" -ti windows/servercore:ltsc2019
> C:\> hosname
> test.company.com
> C:\>(Get-WmiObject win32_computersystem).DNSHostName
> .DNSHostName was unexpected at this time.

We need to yet test this from Kubernetes, but since Kubernetes relies on Docker to set these things, I think it should work? There is still doubts that this will be useful at all for the Windows applications. Today, we discussed this topic during the Sig-Windows, they have been asking around within the Windows community to conclude whether this feature makes sense for them. Since it is an optional field in podspec, we agreed to just make sure we mention in the docs we write about this feature the expected behavior in Windows.

**Comment by javidiaz** [comment]
I intend to close this PR at some point. I split it in two new PRs to separate API changes https://github.com/kubernetes/kubernetes/pull/91697 from Kubelet changes https://github.com/kubernetes/kubernetes/pull/91699

**Comment by javidiaz** [comment]
We will leave all in one PR but use https://github.com/kubernetes/kubernetes/pull/91699 that has commits already organized by meaningful changes.

**Pull request closed without merging**